src/clj/puppetlabs/services/master/file_serving.clj

```diff
@@ -0,0 +1,364 @@
+     (yaml/parse-string (slurp config-path)))))
+
+(def default-project-modulepath
+  ["modules" "site-modules" "site"])
+
+(defn parse-modulepath
+  "The modulepath for a bolt project can either be an array of paths or a string
+  with a path separator."
+  [modulepath]
+  (if (string? modulepath)
+    (str/split modulepath #":")
+    modulepath))
+
+(defn get-project-modulepath
+  "Given a map representing the bolt project configuration, return the
+  modulepath. If the modulepath is not defined or the configuration map passed
+  is nil then return the default modulepath."
+  [project-config]
+  (let [has-modules-key? (contains? project-config :modules)]
+    (if-let [modulepath (parse-modulepath (get project-config :modulepath))]
+      (if has-modules-key?
+        (concat modulepath [".modules"])
+        modulepath)
+      (if has-modules-key?
+        ["modules" ".modules"]
+        default-project-modulepath))))
+
+(defn find-project-file
+  "Find a file in a project using the parameters from a `file_content` request.
+  Returns the path as as string. If the module name is the same as the project
+  name then files are served from the project directly."
+  [bolt-projects-dir project-ref mount module path]
+  (when (is-bolt-project? (str bolt-projects-dir "/" project-ref))
+    (let [project-root (get-project-root bolt-projects-dir project-ref)
+          project-config (read-bolt-project-config project-root)
+          project-name (get project-config :name)
+          modulepath (get-project-modulepath project-config)
+          module-root (if (= project-name module)
+                        project-root
+                        (find-project-module project-root module modulepath))
+          file-path (str module-root "/" (mount->path-component mount) "/" path)]
+      (when (fs/exists? file-path)
+        file-path))))
+
+(defn mount-dirs-in-modulepath
+  [mount modulepath project-root project-as-module?]
+  (let [sub-dir (case mount
+                    "plugins" "lib"
+                    "pluginfacts" "facts.d")
+        modules-in-modulepath (dirs-in-project-modulepath modulepath project-root)
+        module-dirs (if project-as-module?
+                      (cons (fs/file project-root) modules-in-modulepath)
+                      modules-in-modulepath)]
+    (->> module-dirs
+         (map #(.resolve (.toPath %) sub-dir))
+         (filter path-exists?))))
+
+
+(defn plugin-file-if-exists
+  "Given a relative path as received by the plugins mount, and the path of a lib
+  dir in a module, return the path of the file if it exists in the lib dir."
+  [relative-path lib-root]
+  (let [file-path (.resolve lib-root relative-path)]
+    (when (path-exists? file-path)
+      (.toString file-path))))
+
+(defn find-project-plugin-file
+  "Given a relative path as received by the plugins mount, search all lib dirs
+  and return the path to the file if it's found in any of them."
+  [bolt-projects-dir project-ref mount relative-path]
```

Comment:
Is `project-ref` what you're using to refer to name+ref? I think [this email thread](https://groups.google.com/a/puppet.com/d/msgid/team-skeletor/CAOPg%3DQck4EPde9w%2B8gYakbQWdsAAmHtevbLmO_C0L0aVwevqBQ%40mail.gmail.com?utm_medium=email&utm_source=footer) is suggesting we just use `project` for that? `project-ref` makes me think it's _just_ the ref.

Reply:
Yeah, I wrote the initial code before that thead and didn't revisit that issue. I think plain `project` is kind of iffy though so i'm not sure what's best.

Reply:
`project-id-and-ref`? lol I'm not sure either. but that thread was sort of implying to me that `project` means id+ref.

Reply:
I don't think using just `project` makes sense here. Regardless, this is internal and so not exactly subject to the requirements being discussed in that thread. In the context of this code, do we ever deal with name and ref being separate? If not, I'm totally fine leaving this as `project-ref`, because a ref by itself is meaningless (we don't use it, so it's not confusing), but `project-ref` indicates you're passing more than just the name of the project.

Reply:
If we're going to use `project-ref` to indicate both things in one place though, we should use it everywhere to make it clear that's _always_ what it refers to.

Reply:
¯\_(ツ)_/¯ when I see `ref` I always think just git ref, but if y'all wanna leave it, that's fine. I just think we should be as consistent as possible, even if things are internal-only.